lib/Optimizer/Transforms/QuakeAddMetadata.cpp

```diff
@@ -122,6 +124,35 @@ struct QuakeFunctionAnalysis {
                 return WalkResult::interrupt();
               }
             }
+
+            // Look for any future cast/compute_ptr/load, and if that load is
+            // used by a conditional statement
+            if (auto cast = dyn_cast<cudaq::cc::CastOp>(allocUser)) {
+              for (auto castUser : cast->getUsers()) {
+                if (auto cp = dyn_cast<cudaq::cc::ComputePtrOp>(castUser)) {
+                  for (auto cpUser : cp->getUsers()) {
+                    if (auto load = dyn_cast<cudaq::cc::LoadOp>(cpUser)) {
+                      auto loadUser = *load->getUsers().begin();
```

Comment:
There's too much pattern matching on specific trees of IR nodes here, I think. I think we should perform a reaching data-flow analysis over the function to determine what classical operations use the result(s) of a quantum measurement (a transitive closure). That set can then be examined to determine any control operations or control dependent subgraphs of the CFG.

That said, this is ok as a baseline implementation for now. We can open an issue for future improvements.

Reply:
Issue #749 created.